lib/src/main/java/com/segment/analytics/kotlin/destinations/consent/ConsentManager.kt

```diff
@@ -0,0 +1,181 @@
+        if (consentChange == null) {
+            consentChange = { analytics.track(EVENT_SEGMENT_CONSENT_PREFERENCE) }
+        }
+    }
+
+    override fun setup(analytics: Analytics) {
+        super.setup(analytics)
+
+        // Empty state
+        store.provide(ConsentState.defaultState)
+    }
+
+    override fun update(settings: Settings, type: Plugin.UpdateType) {
+
+        val state = consentStateFrom(settings)
+        // Update the store
+        store.dispatch(UpdateConsentStateActionFull(state), ConsentState::class)
+
+
+        analytics?.let {
+
+            // Add Segment Destination blocker
+            val segmentDestination = it.find(Constants.SEGMENT_IO_KEY)
+            segmentDestination.let {
+                val existingBlocker =
+                    segmentDestination?.analytics?.find(SegmentConsentBlocker::class)
+                if (existingBlocker == null) {
+                    segmentDestination?.add(SegmentConsentBlocker(store))
+                }
+            }
+
+            // Add Blocker to all other destinations
+            val destinationKeys = state.destinationCategoryMap.keys
+            for (key in destinationKeys) {
+                val destination = analytics.find(key)
+                destination.let {
+                    if (it?.key != Constants.SEGMENT_IO_KEY) {
+                        val existingBlock =
+                            destination?.analytics?.find(ConsentBlocker::class)
+                        if (existingBlock == null) {
+                            destination?.add(ConsentBlocker(key, store))
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private fun consentStateFrom(settings: Settings): ConsentState {
+
+        val destinationMapping = mutableMapOf<String, Array<String>>()
+        var hasUnmappedDestinations = true
+        var enabledAtSegment = true
+
+        // Add all mappings
+        settings.integrations.forEach { integrationName, integrationJson ->
+            // If the integration has the consent key:
+            integrationJson.jsonObject[CONSENT_SETTINGS_KEY]?.let {
+
+                // Build list of categories required for this integration
+                val categories: MutableList<String> = mutableListOf()
+                (it.jsonObject.get(CATEGORIES_KEY) as JsonArray).forEach { categoryJsonElement ->
+                    categories.add(categoryJsonElement.toString().replace("\"", "").trim())
+                }
+                destinationMapping[integrationName] = categories.toTypedArray()
+            }
+        }
+
+        // Set hasUnmappedDestinations
+        try {
+            settings.toJsonElement().jsonObject.get(CONSENT_SETTINGS_KEY)?.let {
+                val jsonElement = it.jsonObject.get(HAS_UNMAPPED_DESTINATIONS_KEY)
+                println("hasUnmappedDestinations jsonElement: $jsonElement")
+                hasUnmappedDestinations = jsonElement.toString() == "true"
```

Comment:
we have a set of utility method that can simplify operation like this. see [here](https://github.com/segmentio/analytics-kotlin/blob/main/core/src/main/java/com/segment/analytics/kotlin/core/utilities/JSON.kt#L113)
for example:
* get object: `jsonObject.get("s")?.safeJsonObject`
* get string:  `jsonObject.getString("s")`
* get boolean: `jsonObject.getBoolean("s")`
* etc

Reply:
oooh very nice